File: FrozenFramework/src/main/java/lakeshire/github/com/frozenframework/fragment/BasePullFragment.java
package lakeshire.github.com.frozenframework.fragment;

import android.view.View;

import in.srain.cube.views.ptr.PtrFrameLayout;
import in.srain.cube.views.ptr.PtrHandler;
import in.srain.cube.views.ptr.PtrUIHandler;
import in.srain.cube.views.ptr.header.MaterialHeader;
import lakeshire.github.com.frozenframework.R;

/**
 * 支持下拉的Fragment
 *
 * @author lakeshire
 */
public abstract class BasePullFragment extends BaseFragment {
    protected PtrFrameLayout mPtrFrameLayout;

    @Override
    public void initUi() {
        mPtrFrameLayout = (PtrFrameLayout) mContainerView.findViewById(R.id.ptr_frame);
        initPtrFrame();
    }

    protected void initPtrFrame() {
        View header = getHeader();
        mPtrFrameLayout.setDurationToCloseHeader(1500);
        mPtrFrameLayout.setDurationToClose(300);
        mPtrFrameLayout.setResistance(1.5f);
        mPtrFrameLayout.setRatioOfHeaderHeightToRefresh(1.0f);
        mPtrFrameLayout.setKeepHeaderWhenRefresh(true);
        mPtrFrameLayout.setPullToRefresh(true);
        mPtrFrameLayout.setHeaderView(header);
        if (header instanceof PtrUIHandler) {
            mPtrFrameLayout.addPtrUIHandler((PtrUIHandler) header);
        }
        mPtrFrameLayout.setPtrHandler(new PtrHandler() {
            @Override
            public boolean checkCanDoRefresh(PtrFrameLayout frame, View content, View header) {
                return checkCanRefresh(frame, content, header);
            }

            @Override
            public void onRefreshBegin(PtrFrameLayout frame) {
                onRefresh();
            }
        });

    }

    protected View getHeader() {
        MaterialHeader header = new MaterialHeader(mContext);
        header.setPadding(0, 0, 0, 32);
        return header;
    }

    protected void refreshComplete() {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mPtrFrameLayout.refreshComplete();
            }
        });
    }

    abstract protected void onRefresh();

    abstract protected boolean checkCanRefresh(PtrFrameLayout frame, View content, View header);
}

File: FrozenFramework/src/main/java/lakeshire/github/com/frozenframework/fragment/BasePullFragment.kt
package lakeshire.github.com.frozenframework.fragment

import `in`.srain.cube.views.ptr.PtrFrameLayout
import `in`.srain.cube.views.ptr.PtrHandler
import `in`.srain.cube.views.ptr.PtrUIHandler
import `in`.srain.cube.views.ptr.header.MaterialHeader
import android.view.View
import lakeshire.github.com.frozenframework.R

/**
 * 支持下拉的Fragment

 * @author lakeshire
 */
abstract class BasePullFragment : BaseFragment() {
    lateinit var mPtrFrameLayout: PtrFrameLayout

    override fun initUi() {
        mPtrFrameLayout = mContainerView.findViewById<View>(R.id.ptr_frame) as PtrFrameLayout
        initPtrFrame()
    }

    fun initPtrFrame() {
        val header = header
        mPtrFrameLayout.setDurationToCloseHeader(1500)
        mPtrFrameLayout.setDurationToClose(300)
        mPtrFrameLayout.resistance = 1.5f
        mPtrFrameLayout.setRatioOfHeaderHeightToRefresh(1.0f)
        mPtrFrameLayout.isKeepHeaderWhenRefresh = true
        mPtrFrameLayout.isPullToRefresh = true
        mPtrFrameLayout.headerView = header
        if (header is PtrUIHandler) {
            mPtrFrameLayout.addPtrUIHandler(header as PtrUIHandler)
        }
        mPtrFrameLayout.setPtrHandler(object : PtrHandler {
            override fun checkCanDoRefresh(frame: PtrFrameLayout, content: View, header: View): Boolean {
                return checkCanRefresh(frame, content, header)
            }

            override fun onRefreshBegin(frame: PtrFrameLayout) {
                onRefresh()
            }
        })

    }

    val header: View
        get() {
            val header = MaterialHeader(mContext)
            header.setPadding(0, 0, 0, 32)
            return header
        }

    fun refreshComplete() {
        runOnUiThread(Runnable { mPtrFrameLayout.refreshComplete() })
    }

    abstract fun onRefresh()

    abstract fun checkCanRefresh(frame: PtrFrameLayout, content: View, header: View): Boolean
}
